File: parser/src/identity.ts
/** ALP Self-Sovereign Identity (v18.0.0 — V14 The Sovereign Era).
 *
 * W3C DID-based agent identity without a central authority:
 *
 * - `AgentIdentity`         — creates/manages DIDs (decentralized identifiers).
 * - `IdentityResolver`      — verifies presentations against a trust registry.
 * - `TrustRegistry`         — maps DIDs to permission scopes and trust levels.
 * - `VerifiablePresentation` — signed identity proof from agent to verifier.
 * - `AgentKeyStore`         — persists agent key pairs for DID operations.
 *
 * Mirrors `sdk/python/alp_sdk/identity.py`.
 */

import * as crypto from 'node:crypto'
import * as fs from 'fs'
import * as path from 'path'

export interface AgentIdentityData {
  did: string
  agent_id: string
  public_key: string
  created_at: string
  metadata: Record<string, any>
}

export class AgentIdentity {
  did: string
  agent_id: string
  public_key: string
  created_at: string
  metadata: Record<string, any>

  constructor(did: string, agent_id: string, public_key: string, created_at = '', metadata: Record<string, any> = {}) {
    this.did = did
    this.agent_id = agent_id
    this.public_key = public_key
    this.created_at = created_at || new Date().toISOString()
    this.metadata = metadata
  }

  toDict(): AgentIdentityData {
    return {
      did: this.did,
      agent_id: this.agent_id,
      public_key: this.public_key,
      created_at: this.created_at,
      metadata: this.metadata,
    }
  }

  static fromDict(d: Record<string, any>): AgentIdentity {
    return new AgentIdentity(
      d.did ?? d['did'],
      d.agent_id ?? d['agent_id'],
      d.public_key ?? d['public_key'],
      d.created_at ?? '',
      d.metadata ?? {},
    )
  }
}

export interface VerifiablePresentationData {
  did: string
  agent_id: string
  claims: Record<string, any>
  signature: string
  issued_at: string
}

export class VerifiablePresentation {
  did: string
  agent_id: string
  claims: Record<string, any>
  signature: string
  issued_at: string

  constructor(did: string, agent_id: string, claims: Record<string, any>, signature: string, issued_at = '') {
    this.did = did
    this.agent_id = agent_id
    this.claims = claims
    this.signature = signature
    this.issued_at = issued_at || new Date().toISOString()
  }

  toDict(): VerifiablePresentationData {
    return {
      did: this.did,
      agent_id: this.agent_id,
      claims: this.claims,
      signature: this.signature,
      issued_at: this.issued_at,
    }
  }

  verify(public_key: string): boolean {
    const payload = JSON.stringify({ did: this.did, agent_id: this.agent_id, claims: this.claims })
    const expected = sha256(payload + public_key)
    return this.signature === expected
  }
}

export interface TrustEntry {
  agent_id: string
  scopes: string[]
  trust_level: string
  registered_at: string
}

export class TrustRegistry {
  private alp_dir: string
  private entries: Record<string, TrustEntry> = {}

  constructor(alp_dir: string) {
    this.alp_dir = alp_dir
    this.load()
  }

  private identityDir(): string {
    return path.join(this.alp_dir, '.identity')
  }

  private trustPath(): string {
    return path.join(this.identityDir(), 'trust_registry.json')
  }

  load(): void {
    const p = this.trustPath()
    if (!fs.existsSync(p)) return
    try {
      const data = JSON.parse(fs.readFileSync(p, 'utf-8'))
      if (typeof data === 'object' && data) {
        this.entries = data as Record<string, TrustEntry>
      }
    } catch {
      this.entries = {}
    }
  }

  save(): void {
    const d = this.identityDir()
    if (!fs.existsSync(d)) {
      fs.mkdirSync(d, { recursive: true })
    }
    fs.writeFileSync(this.trustPath(), JSON.stringify(this.entries, null, 2))
  }

  register(did: string, agent_id: string, scopes: string[], trust_level = 'standard'): TrustEntry {
    const entry: TrustEntry = {
      agent_id,
      scopes,
      trust_level,
      registered_at: new Date().toISOString(),
    }
    this.entries[did] = entry
    this.save()
    return entry
  }

  resolve(did: string): TrustEntry | undefined {
    return this.entries[did]
  }

  revoke(did: string): boolean {
    if (did in this.entries) {
      delete this.entries[did]
      this.save()
      return true
    }
    return false
  }

  listDids(): string[] {
    return Object.keys(this.entries)
  }

  hasScope(did: string, required_scope: string): boolean {
    const entry = this.entries[did]
    if (!entry) return false
    return entry.scopes.includes(required_scope)
  }
}

export class IdentityResolver {
  constructor(private trust_registry: TrustRegistry) {}

  verifyPresentation(presentation: VerifiablePresentation, public_key: string): Record<string, any> {
    if (!presentation.verify(public_key)) {
      return { valid: false, reason: 'invalid_signature' }
    }
    const entry = this.trust_registry.resolve(presentation.did)
    if (!entry) {
      return { valid: false, reason: 'unknown_did' }
    }
    return {
      valid: true,
      did: presentation.did,
      agent_id: presentation.agent_id,
      scopes: entry.scopes,
      trust_level: entry.trust_level,
    }
  }
}

export interface KeyPair {
  public_key: string
  private_key: string
}

export class AgentKeyStore {
  private alp_dir: string
  private keys: Record<string, KeyPair> = {}

  constructor(alp_dir: string) {
    this.alp_dir = alp_dir
    this.load()
  }

  private identityDir(): string {
    return path.join(this.alp_dir, '.identity')
  }

  private keysPath(): string {
    return path.join(this.identityDir(), 'agent_keys.json')
  }

  load(): void {
    const p = this.keysPath()
    if (!fs.existsSync(p)) return
    try {
      this.keys = JSON.parse(fs.readFileSync(p, 'utf-8'))
    } catch {
      this.keys = {}
    }
  }

  save(): void {
    const d = this.identityDir()
    if (!fs.existsSync(d)) {
      fs.mkdirSync(d, { recursive: true })
    }
    fs.writeFileSync(this.keysPath(), JSON.stringify(this.keys, null, 2))
  }

  storeKey(did: string, public_key: string, private_key: string): void {
    this.keys[did] = { public_key, private_key }
    this.save()
  }

  getKey(did: string): KeyPair | undefined {
    return this.keys[did]
  }

  removeKey(did: string): boolean {
    if (did in this.keys) {
      delete this.keys[did]
      this.save()
      return true
    }
    return false
  }
}

export function generateKeypair(): KeyPair {
  const private_key = crypto.randomUUID().replace(/-/g, '')
  const public_key = sha256(private_key)
  return { public_key, private_key }
}

export function createDid(agent_id: string, public_key: string): string {
  const key_hash = sha256(public_key).slice(0, 16)
  return `did:alp:${agent_id}:${key_hash}`
}

function sha256(input: string): string {
  return crypto.createHash('sha256').update(input).digest('hex')
}

File: sdk/python/alp_sdk/identity.py
"""ALP Self-Sovereign Identity (v18.0.0 — V14 The Sovereign Era).

W3C DID-based agent identity without a central authority:

* ``AgentIdentity``   — creates/manages DIDs (decentralized identifiers).
* ``IdentityResolver`` — verifies presentations against a trust registry.
* ``TrustRegistry``    — maps DIDs to permission scopes and trust levels.
* ``VerifiablePresentation`` — signed identity proof from agent to verifier.

Mirrors the planned ``parser/src/identity.ts`` surface; tests live in
``sdk/python/tests/test_identity.py``.
"""

import hashlib
import json
import os
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Set


IDENTITY_DIR = ".identity"
TRUST_FILE = "trust_registry.json"
KEYS_FILE = "agent_keys.json"


def identity_dir(alp_dir: str) -> str:
    return os.path.join(alp_dir, IDENTITY_DIR)


def trust_path(alp_dir: str) -> str:
    return os.path.join(identity_dir(alp_dir), TRUST_FILE)


def keys_path(alp_dir: str) -> str:
    return os.path.join(identity_dir(alp_dir), KEYS_FILE)


@dataclass
class AgentIdentity:
    """W3C DID-based agent identity."""

    did: str
    agent_id: str
    public_key: str
    created_at: str = ""
    metadata: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self):
        if not self.created_at:
            self.created_at = _now_iso()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "did": self.did,
            "agent_id": self.agent_id,
            "public_key": self.public_key,
            "created_at": self.created_at,
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "AgentIdentity":
        return cls(
            did=d["did"],
            agent_id=d["agent_id"],
            public_key=d["public_key"],
            created_at=d.get("created_at", ""),
            metadata=d.get("metadata", {}),
        )


@dataclass
class VerifiablePresentation:
    """Signed identity proof from agent to verifier."""

    did: str
    agent_id: str
    claims: Dict[str, Any]
    signature: str
    issued_at: str = ""

    def __post_init__(self):
        if not self.issued_at:
            self.issued_at = _now_iso()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "did": self.did,
            "agent_id": self.agent_id,
            "claims": self.claims,
            "signature": self.signature,
            "issued_at": self.issued_at,
        }

    def verify(self, public_key: str) -> bool:
        payload = json.dumps({"did": self.did, "agent_id": self.agent_id, "claims": self.claims}, sort_keys=True).encode()
        expected = hashlib.sha256(payload + public_key.encode()).hexdigest()
        return self.signature == expected


class TrustRegistry:
    """Maps DIDs to permission scopes and trust levels."""

    def __init__(self, alp_dir: str):
        self.alp_dir = alp_dir
        self._entries: Dict[str, Dict[str, Any]] = {}
        self._load()

    def _load(self) -> None:
        p = trust_path(self.alp_dir)
        if not os.path.exists(p):
            return
        try:
            with open(p, "r", encoding="utf-8") as f:
                data = json.load(f)
            if isinstance(data, dict):
                self._entries = data
        except Exception:
            self._entries = {}

    def _save(self) -> None:
        d = identity_dir(self.alp_dir)
        if not os.path.exists(d):
            os.makedirs(d, exist_ok=True)
        with open(trust_path(self.alp_dir), "w", encoding="utf-8") as f:
            json.dump(self._entries, f, indent=2)

    def register(self, did: str, agent_id: str, scopes: List[str], trust_level: str = "standard") -> Dict[str, Any]:
        entry = {
            "agent_id": agent_id,
            "scopes": scopes,
            "trust_level": trust_level,
            "registered_at": _now_iso(),
        }
        self._entries[did] = entry
        self._save()
        return entry

    def resolve(self, did: str) -> Optional[Dict[str, Any]]:
        return self._entries.get(did)

    def revoke(self, did: str) -> bool:
        if did in self._entries:
            del self._entries[did]
            self._save()
            return True
        return False

    def list_dids(self) -> List[str]:
        return list(self._entries.keys())

    def has_scope(self, did: str, required_scope: str) -> bool:
        entry = self._entries.get(did)
        if not entry:
            return False
        return required_scope in entry.get("scopes", [])


class IdentityResolver:
    """Verifies presentations against a trust registry."""

    def __init__(self, trust_registry: TrustRegistry):
        self.trust_registry = trust_registry

    def verify_presentation(self, presentation: VerifiablePresentation, public_key: str) -> Dict[str, Any]:
        if not presentation.verify(public_key):
            return {"valid": False, "reason": "invalid_signature"}
        entry = self.trust_registry.resolve(presentation.did)
        if not entry:
            return {"valid": False, "reason": "unknown_did"}
        return {
            "valid": True,
            "did": presentation.did,
            "agent_id": presentation.agent_id,
            "scopes": entry.get("scopes", []),
            "trust_level": entry.get("trust_level", "standard"),
        }


class AgentKeyStore:
    """Persists agent key pairs for DID operations."""

    def __init__(self, alp_dir: str):
        self.alp_dir = alp_dir
        self._keys: Dict[str, Dict[str, str]] = {}
        self._load()

    def _load(self) -> None:
        p = keys_path(self.alp_dir)
        if not os.path.exists(p):
            return
        try:
            with open(p, "r", encoding="utf-8") as f:
                self._keys = json.load(f)
        except Exception:
            self._keys = {}

    def _save(self) -> None:
        d = identity_dir(self.alp_dir)
        if not os.path.exists(d):
            os.makedirs(d, exist_ok=True)
        with open(keys_path(self.alp_dir), "w", encoding="utf-8") as f:
            json.dump(self._keys, f, indent=2)

    def store_key(self, did: str, public_key: str, private_key: str) -> None:
        self._keys[did] = {"public_key": public_key, "private_key": private_key}
        self._save()

    def get_key(self, did: str) -> Optional[Dict[str, str]]:
        return self._keys.get(did)

    def remove_key(self, did: str) -> bool:
        if did in self._keys:
            del self._keys[did]
            self._save()
            return True
        return False


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()


def generate_keypair() -> tuple[str, str]:
    private_key = uuid.uuid4().hex
    public_key = hashlib.sha256(private_key.encode()).hexdigest()
    return public_key, private_key


def create_did(agent_id: str, public_key: str) -> str:
    key_hash = hashlib.sha256(public_key.encode()).hexdigest()[:16]
    return f"did:alp:{agent_id}:{key_hash}"
